File: Harvest/Services/AuthService.swift
import Foundation
import Supabase
import Auth

struct AuthService {
    private var client: SupabaseClient { SupabaseManager.shared.client }

    func signUp(email: String, password: String) async throws -> Auth.User {
        let response = try await client.auth.signUp(
            email: email,
            password: password,
            data: ["app_name": .string("Harvest")],
            redirectTo: URL(string: "\(Config.appScheme)://auth/callback")
        )
        return response.user
    }

    func signIn(email: String, password: String) async throws -> Auth.Session {
        let session = try await client.auth.signIn(
            email: email,
            password: password
        )
        return session
    }

    func signOut() async throws {
        try await client.auth.signOut()
    }

    func deleteAccount(userId: String) async throws {
        struct DeleteAccountResponse: Decodable {
            let success: Bool
            let message: String?
        }

        let session = try await client.auth.session
        let endpoint = Config.supabaseURL
            .appendingPathComponent("functions")
            .appendingPathComponent("v1")
            .appendingPathComponent("delete-account")

        var request = URLRequest(url: endpoint)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(Config.supabaseAnonKey, forHTTPHeaderField: "apikey")
        request.setValue("Bearer \(session.accessToken)", forHTTPHeaderField: "Authorization")
        request.httpBody = try JSONSerialization.data(withJSONObject: ["user_id": userId])

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw NSError(
                domain: "AuthService",
                code: 1002,
                userInfo: [NSLocalizedDescriptionKey: "Account deletion failed because the backend response was invalid."]
            )
        }

        guard 200..<300 ~= httpResponse.statusCode else {
            let backendMessage = String(data: data, encoding: .utf8)?.trimmingCharacters(in: .whitespacesAndNewlines)
            let friendlyMessage: String

            if httpResponse.statusCode == 404 {
                friendlyMessage = "Account deletion is not available yet because the delete-account Edge Function has not been deployed."
            } else if let backendMessage, !backendMessage.isEmpty {
                friendlyMessage = backendMessage
            } else {
                friendlyMessage = "Account deletion failed. Please try again or contact support."
            }

            throw NSError(
                domain: "AuthService",
                code: httpResponse.statusCode,
                userInfo: [NSLocalizedDescriptionKey: friendlyMessage]
            )
        }

        if let decoded = try? JSONDecoder().decode(DeleteAccountResponse.self, from: data),
           decoded.success == false {
            throw NSError(
                domain: "AuthService",
                code: 1003,
                userInfo: [NSLocalizedDescriptionKey: decoded.message ?? "Account deletion failed."]
            )
        }
    }

    func getCurrentSession() async throws -> Auth.Session? {
        try await client.auth.session
    }

    func getCurrentUser() async throws -> Auth.User? {
        try await client.auth.session.user
    }

    func authStateChanges() -> AsyncStream<(AuthChangeEvent, Auth.Session?)> {
        AsyncStream { continuation in
            let task = Task {
                for await (event, session) in client.auth.authStateChanges {
                    continuation.yield((event, session))
                }
                continuation.finish()
            }
            continuation.onTermination = { _ in
                task.cancel()
            }
        }
    }
}

enum AuthError: LocalizedError {
    case signUpFailed
    case notAuthenticated

    var errorDescription: String? {
        switch self {
        case .signUpFailed: return "Failed to create account"
        case .notAuthenticated: return "Not authenticated"
        }
    }
}

File: android/app/src/main/java/com/harvestglass/harvest/data/service/AuthService.kt
package com.harvestglass.harvest.data.service

import com.harvestglass.harvest.data.model.UserProfile
import com.harvestglass.harvest.Config
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.network.SupabaseHttpClient
import io.ktor.client.request.header
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.client.statement.bodyAsText
import io.ktor.http.ContentType
import io.ktor.http.HttpStatusCode
import io.ktor.http.contentType
import io.ktor.http.isSuccess
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import io.github.jan.supabase.auth.auth
import io.github.jan.supabase.auth.providers.builtin.Email
import io.github.jan.supabase.auth.status.SessionStatus
import io.github.jan.supabase.postgrest.postgrest
import kotlinx.coroutines.flow.Flow

/** Mirrors Harvest/Services/AuthService.swift. */
class AuthService(private val client: SupabaseClient) {

    /**
     * iOS uses `session.user.id.uuidString.lowercased()` everywhere. RLS
     * comparisons depend on the lowercase form, so normalise on the way out.
     */
    private fun String.normalizedId() = lowercase()

    suspend fun signUp(email: String, password: String): String? {
        val user = client.auth.signUpWith(Email) {
            this.email = email
            this.password = password
        }
        return user?.id?.normalizedId()
    }

    suspend fun signIn(email: String, password: String): String {
        client.auth.signInWith(Email) {
            this.email = email
            this.password = password
        }
        return requireNotNull(currentUserId()) { "Sign-in returned no session" }
    }

    suspend fun signOut() = client.auth.signOut()

    /**
     * Deletes the account via the `delete-account` Edge Function.
     *
     * Mirrors AuthService.swift, including its friendlier messages: a 404
     * means the function is not deployed, which is a different problem from a
     * failed deletion and should read that way.
     */
    suspend fun deleteAccount(userId: String) {
        val session = client.auth.currentSessionOrNull()
            ?: throw IllegalStateException("Not authenticated")

        val response = client.httpClient.post(
            "${Config.SUPABASE_URL}/functions/v1/delete-account"
        ) {
            header("apikey", Config.SUPABASE_ANON_KEY)
            header("Authorization", "Bearer ${session.accessToken}")
            contentType(ContentType.Application.Json)
            setBody(buildJsonObject { put("user_id", userId) })
        }

        if (response.status.isSuccess()) return

        val body = runCatching { response.bodyAsText() }.getOrNull()?.trim()
        throw IllegalStateException(
            when {
                response.status == HttpStatusCode.NotFound ->
                    "Account deletion is not available yet because the delete-account " +
                        "Edge Function has not been deployed."
                !body.isNullOrEmpty() -> body
                else -> "Account deletion failed. Please try again or contact support."
            }
        )
    }

    fun currentUserIdOrNull(): String? = client.auth.currentUserOrNull()?.id?.normalizedId()

    suspend fun currentUserId(): String? = currentUserIdOrNull()

    fun sessionStatus(): Flow<SessionStatus> = client.auth.sessionStatus

    suspend fun loadProfile(userId: String): UserProfile? =
        client.postgrest.from("users")
            .select { filter { eq("id", userId) } }
            .decodeSingleOrNull<UserProfile>()
}
